zcash_client_backend/src/tor/http/exchange_rate.rs

```diff
@@ -0,0 +1,373 @@
+
+impl Client {
+    /// Fetches the latest `currency`/ZEC exchange rate, derived from several exchanges.
+    ///
+    /// Supported currencies:
+    /// - USD
+    ///
+    /// Returns:
+    /// - `Ok(Some(rate))` if at least one exchange request succeeds.
+    /// - `Ok(None)` if the given currency is unsupported.
+    /// - `Err(_)` if none of the exchange queries succeed.
+    pub async fn get_exchange_rate(&self, currency: Currency) -> Result<Option<Decimal>, Error> {
+        let pair = match ExchangePair::get(currency) {
+            Some(pair) => pair,
+            None => return Ok(None),
+        };
+        assert!(matches!(pair, ExchangePair::Usd));
+
+        // Fetch the data in parallel.
+        let res = join!(
+            Binance::query_usd(self),
+            Coinbase::query_usd(self),
+            GateIo::query_usd(self),
+            Gemini::query_usd(self),
+            KuCoin::query_usd(self),
+            Mexc::query_usd(self),
+        );
+        trace!(?res, "Data results");
+        let (binance, coinbase, gate_io, gemini, ku_coin, mexc) = res;
+
+        // Split into successful queries and errors.
+        fn split<T: ExchangeData>(s: &mut Vec<Decimal>, e: &mut Vec<Error>, res: Result<T, Error>) {
+            match res {
+                Ok(d) => s.push(d.price()),
+                Err(error) => e.push(error),
+            }
+        }
+        let mut prices = vec![];
+        let mut errors = vec![];
+        split(&mut prices, &mut errors, binance);
+        split(&mut prices, &mut errors, coinbase);
+        split(&mut prices, &mut errors, gate_io);
+        // We handle Gemini below to exclude it from eviction.
+        split(&mut prices, &mut errors, ku_coin);
+        split(&mut prices, &mut errors, mexc);
+
+        // "Never go to sea with two chronometers; take one or three."
+        // Randomly drop one price if necessary to have an odd number of prices.
+        let evict_random = |s: &mut Vec<Decimal>| {
+            if let Some(index) = (0..s.len()).choose(&mut thread_rng()) {
+                s.remove(index);
+            }
+        };
+        if let Ok(gemini) = gemini {
+            if prices.len() % 2 != 0 {
+                evict_random(&mut prices);
+            }
+            prices.push(gemini.price());
+        } else {
+            if prices.len() % 2 == 0 {
+                evict_random(&mut prices);
+            }
+        };
+
+        // If all of the requests failed, log all errors and return one of them.
+        if prices.is_empty() {
+            error!("All exchange requests failed");
+            Err(errors.into_iter().next().expect("All requests failed"))
+        } else {
+            // We have an odd number of prices; take the median.
+            assert!(prices.len() % 2 != 0);
+            prices.sort();
+            let median = prices.len() / 2;
+            Ok(Some(prices[median]))
+        }
+    }
+}
+
+#[derive(Clone, Copy)]
+enum ExchangePair {
```

Comment:
I'm skeptical of this type. Why isn't it effectively a tuple of `Currency` values, so that it can represent both halves of the exchange rate? `ExchangePair` is a misleading name for it.

Reply:
This type is left over from when I mistakenly though that there would be _any_ exchange data sources that reliably gave ZEC pairs in anything other than USD(T). It is deleted if I make the change to `get_latest_zec_to_usd_rate` I suggested above.

Reply:
Also, we _do not want_ an API that allows for arbitrary cryptocurrency exchange rate lookups; this API is specifically for looking up ZEC/`something` pairs, and if I make the change per above, then specifically ZEC/USD pairs.